File: app/src/main/java/com/mss/imagesearcher/di/AppModule.java
package com.mss.imagesearcher.di;

import android.content.Context;
import android.support.annotation.NonNull;

import com.mss.imagesearcher.domain.ImageListModel;
import com.mss.imagesearcher.domain.repositories.CacheDBRepository;
import com.mss.imagesearcher.domain.repositories.FilesRepository;
import com.mss.imagesearcher.domain.repositories.ImagesNetRepository;
import com.mss.imagesearcher.presenter.imagelist.ImageListPresenter;
import com.mss.imagesearcher.view.helpers.ImageLoader;
import com.mss.imagesearcher.view.helpers.ImageLoaderImpl;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class AppModule {

    private final Context appContext;

    public AppModule(@NonNull final Context appContext) {
        this.appContext = appContext;
    }

    @Singleton
    @Provides
    @NonNull
    Context provideAppContext() {
        return appContext;
    }

    @Singleton
    @Provides
    @NonNull
    ImageListModel provideImageListModel(@NonNull final CacheDBRepository cacheDBRepository,
                                         @NonNull final ImagesNetRepository imagesNetRepository,
                                         @NonNull final FilesRepository filesRepository) {
        return new ImageListModel(cacheDBRepository, imagesNetRepository, filesRepository);
    }

    @Singleton
    @Provides
    @NonNull
    ImageListPresenter provideImageListPresenter(@NonNull final ImageListModel model) {
        return new ImageListPresenter(model);
    }

    @Singleton
    @Provides
    @NonNull
    ImageLoader provideImageLoader() {
        return new ImageLoaderImpl();
    }
}

File: app/src/main/java/com/mss/imagesearcher/di/AppModule.kt
package com.mss.imagesearcher.di

import android.content.Context
import com.mss.imagesearcher.model.ImageListModel
import com.mss.imagesearcher.model.repositories.DBRepository
import com.mss.imagesearcher.model.repositories.FilesRepository
import com.mss.imagesearcher.model.repositories.ImagesNetRepository
import com.mss.imagesearcher.presenter.history.HistoryPresenter
import com.mss.imagesearcher.presenter.imagelist.ImageListPresenter
import com.mss.imagesearcher.presenter.info.InfoPresenter
import com.mss.imagesearcher.presenter.main.MainPresenter
import com.mss.imagesearcher.presenter.settings.SettingsPresenter
import com.mss.imagesearcher.view.helpers.ImageLoader
import com.mss.imagesearcher.view.helpers.ImageLoaderImpl
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class AppModule(private val appContext: Context) {

    @Singleton
    @Provides
    fun provideAppContext(): Context {
        return appContext
    }

    @Singleton
    @Provides
    fun provideImageListModel(//@NonNull final CacheDBRepository cacheDBRepository,
            imagesNetRepository: ImagesNetRepository,
            filesRepository: FilesRepository,
            dbRepository: DBRepository): ImageListModel {
        return ImageListModel(//cacheDBRepository,
                imagesNetRepository,
                filesRepository,
                dbRepository)
    }

    @Singleton
    @Provides
    fun provideMainPresenter(model: ImageListModel): MainPresenter {
        return MainPresenter(model)
    }

    @Singleton
    @Provides
    fun provideHistoryPresenter(model: ImageListModel): HistoryPresenter {
        return HistoryPresenter(model)
    }

    @Singleton
    @Provides
    fun provideSettingsPresenter(model: ImageListModel): SettingsPresenter {
        return SettingsPresenter(model)
    }

    @Singleton
    @Provides
    fun provideImageListPresenter(model: ImageListModel): ImageListPresenter {
        return ImageListPresenter(model)
    }

    @Singleton
    @Provides
    fun provideInfoPresenter(model: ImageListModel): InfoPresenter {
        return InfoPresenter(model)
    }

    @Singleton
    @Provides
    fun provideImageLoader(): ImageLoader {
        return ImageLoaderImpl()
    }
}
